Add magrittr 2.0 blog post (#463)
tidyverse/tidyverse.org: content/blog/magrittr-2-0/index.Rmd

---
output: hugodown::hugo_document

slug: magrittr-2-0
title: magrittr 2.0 is coming soon
date: 2020-07-30
author: Lionel Henry

> description: >

Reviewer: Needs a description

    A 2-3 sentence description of the post that appears on the articles page.
    This can be omitted if it would just recapitulate the title.

photo:
  url: https://unsplash.com/photos/X-NAMq6uP3Q
  author: Mike Benna

categories: [package]
tags: []
---

```{r, echo = FALSE, message = FALSE}
knitr::opts_chunk$set(collapse = T, comment = "#>")
```

> It is with unclouded exhilaration that we announce the impending release of [magrittr](https://magrittr.tidyverse.org/) 2.0. magrittr is the package home to the `%>%` pipe operator written by Stefan Milton Bache and used throughout the tidyverse.

Reviewer: ```suggestion
It is with unclouded exhilaration that we announce the upcoming release of [magrittr](https://magrittr.tidyverse.org/) 2.0. magrittr is the package home to the `%>%` pipe operator written by Stefan Milton Bache and used throughout the tidyverse.
```

> With this major release, we intend to bring the behaviour of the pipe closer to the base pipe `|>` that will likely be included in a future version of R. We also fixed some longstanding issues that made the pipe problematic for programming in packages.

Reviewer: Needs an explicit call for action. Something like:

> This a major rewrite which adds laziness, gives better backtraces, and has much improved performance. Our testing suggests that it should be a drop in replacement, but we'd really like you to try it out and give us some feedback before we submit to CRAN. You can try the development version from GitHub with:

...

> If you discover any issues, please let us know, by ...

You can try this development version from github with:

```{r, eval = FALSE}
# install.packages("remotes")
remotes::install_github("magrittr")
```

This blog post covers the three main changes in this new version of the magrittr pipe.

```{r setup}
library(magrittr)
```

## Laziness

The main user-visible change in this release is that the pipe expressions are now evaluated lazily, only when needed.

```{r}
ignore_arguments <- function(...) "value"

stop("foo") %>% ignore_arguments()
```

> This has subtle implications but should be mostly backward compatible with existing code. The main source of behaviour change is that some code that previously failed may stop failing if the latter part of the pipeline specifically handle the error.

Reviewer: ```suggestion
This has subtle implications but should be backward compatible with existing pipelines that run without error. The main source of behaviour change is that some code that previously failed may stop failing if the latter part of the pipeline specifically handled the error.
```

```

Thanks to this change, you will now be able to pipe into testthat error expectations, for instance:

```{r}
library(testthat) %>%
  suppressMessages()

{ 1 + "a" } %>%
  expect_error("non-numeric argument")
```

## Backtraces

The R implementation of the magrittr pipe was rather costly in terms of backtrace clutter. This made it difficult to debug errors with functions using the pipe:

```{r, eval = FALSE}
foo <- function() bar()
bar <- function() 1 %>% identity() %>% baz()
baz <- function(x) rlang::abort("oh no")

foo()
#> Error: oh no

rlang::last_trace()
#> <error/rlang_error>
#> oh no
#> Backtrace:
#>      █
#>   1. └─global::foo()
#>   2.   └─global::bar()
#>   3.     └─1 %>% identity() %>% baz()
#>   4.       ├─base::withVisible(eval(quote(`_fseq`(`_lhs`)), env, env))
#>   5.       └─base::eval(quote(`_fseq`(`_lhs`)), env, env)
#>   6.         └─base::eval(quote(`_fseq`(`_lhs`)), env, env)
#>   7.           └─`_fseq`(`_lhs`)
#>   8.             └─magrittr::freduce(value, `_function_list`)
#>   9.               ├─base::withVisible(function_list[[k]](value))
#>  10.               └─function_list[[k]](value)
#>  11.                 └─global::baz(.)
```

> This clutter is now completely fixed:

Reviewer: ```suggestion
This clutter is now completely resolved:
```

```{r, eval = FALSE}
foo()
#> Error: oh no

rlang::last_trace()
#> <error/rlang_error>
#> oh no
#> Backtrace:
#>     █
#>  1. ├─global::foo()
#>  2. │ └─global::bar()
#>  3. │   └─1 %>% identity() %>% baz()
#>  4. └─global::baz(.)
```

> Note however that one consequence of having a lazy pipe is that the whole pipeline is pushed on the stack before errors have a chance to be thrown.

Reviewer: ```suggestion
Note that one consequence of having a lazy pipe is that the whole pipeline will shown on the call stack before any errors are thrown:
```

```{r, eval = FALSE}
f1 <- function(x) x
f2 <- function(x) x
f3 <- function(x) x
f4 <- function(x) x

stop("oh no") %>% f1() %>% f2() %>% f3() %>% f4()
#> Error in f1(.) : oh no

rlang::last_trace()
#> <error/rlang_error>
#> oh no
#> Backtrace:
#>     █
#>  1. ├─stop("oh no") %>% f1() %>% f2() %>% f3() %>% f4()
#>  2. ├─global::f4(.)
#>  3. ├─global::f3(.)
#>  4. ├─global::f2(.)
#>  5. └─global::f1(.)
```

The last function of the pipeline is `f4()`, so that's the first one to be run. It evaluates its argument which is provided by `f3()`, so that's the second function pushed on the stack. And so on until `f1()` needs the result of `stop("oh no")` which causes an error.

## Speed

The pipe is now written in C. This greatly improves the performance. Here is a benchmark for the old R implementation:

```{r, eval = FALSE}
bench::mark(

>   `0` = f1(NULL),

Reviewer: I'd suggest removing this benchmark, since it's orthogonal to the comparison of interest here (i.e. new vs old, not no pipe vs pipe).

  `1` = NULL %>% f1(),
  `2` = NULL %>% f1() %>% f2(),
  `3` = NULL %>% f1() %>% f2() %>% f3(),
  `4` = NULL %>% f1() %>% f2() %>% f3() %>% f4(),
)
#> + + + + + + # A tibble: 5 x 13
#>   expression     min  median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc total_time result
#>   <bch:expr> <bch:t> <bch:t>     <dbl> <bch:byt>    <dbl> <int> <dbl>   <bch:tm> <list>
#> 1 0            258ns   319ns  2808344.        0B      0   10000     0     3.56ms <NULL>
#> 2 1           59.4µs  68.9µs    13648.      280B     59.1  6004    26   439.91ms <NULL>
#> 3 2           82.6µs 101.6µs     9252.      280B     42.8  3894    18   420.87ms <NULL>
#> 4 3          106.4µs 124.7µs     7693.      280B     18.8  3690     9   479.64ms <NULL>
#> 5 4          130.9µs 156.1µs     6173.      280B     18.8  2956     9   478.84ms <NULL>
#> # … with 3 more variables: memory <list>, time <list>, gc <list>
```

> The new C implementation is less costly per pipe expression:

Reviewer: Maybe say something like around 10x faster?

```{r, eval = FALSE}
bench::mark(
  `0` = f1(NULL),
  `1` = NULL %>% f1(),
  `2` = NULL %>% f1() %>% f2(),
  `3` = NULL %>% f1() %>% f2() %>% f3(),
  `4` = NULL %>% f1() %>% f2() %>% f3() %>% f4(),
)
#>   expression      min   median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc total_time result
#>   <bch:expr> <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl> <int> <dbl>   <bch:tm> <list>
#> 1 0             270ns    383ns  2240689.        0B    224.   9999     1     4.46ms <NULL>
#> 2 1            4.47µs   5.95µs   159655.        0B     79.9  9995     5     62.6ms <NULL>
#> 3 2            5.97µs    8.8µs   109534.        0B     32.9  9997     3    91.27ms <NULL>
#> 4 3            8.83µs  10.63µs    89902.        0B     27.0  9997     3    111.2ms <NULL>
#> 5 4           10.99µs  13.18µs    72330.        0B     36.2  9995     5   138.19ms <NULL>
```

> 

Reviewer: We don't generally except this to have much impact on data analysis code, but it might yield meaningful speed ups if you had a pipe inside a tight loop.

---
output: hugodown::hugo_document

slug: magrittr-2-0
title: magrittr 2.0 is coming soon
date: 2020-08-20
author: Lionel Henry
description: >
    A new version of the magrittr package brings laziness, better performance, and leaner backtraces for debugging errors.

photo:
  url: https://unsplash.com/photos/X-NAMq6uP3Q
  author: Mike Benna

categories: [package]
tags: []
---

```{r, echo = FALSE, message = FALSE}
knitr::opts_chunk$set(collapse = T, comment = "#>")
```

It is with unclouded composure that we announce the upcoming release of [magrittr](https://magrittr.tidyverse.org/) 2.0. magrittr is the package home to the `%>%` pipe operator written by Stefan Milton Bache and used throughout the tidyverse.

> This last and final version of magrittr resolves the longstanding issues of overhead and backtrace footprint. It also makes the magrittr pipe more compatible with the likely implementation of the native pipe that will probably be included in the next version of R.

Reviewer: ```suggestion
This last and likely final version of magrittr resolves the longstanding issues of overhead and backtrace footprint. It also makes the magrittr pipe more compatible with a native pipe that will probably be included in the next version of R.
```

> This version of magrittr has been completely rewritten in C to give better backtraces and much improved performance. It also uses a different approach in order to support laziness like the future base pipe will. Our testing suggests that it should be a drop in replacement, but we'd really like you to try it out and give us some feedback before we submit to CRAN. You can install the development version from GitHub with:

Reviewer: ```suggestion
This version of magrittr has been completely rewritten in C to give better backtraces and much improved performance. It also uses a different approach in order to support laziness: this enables new uses of the pipe, and ensures magrittr is as similar as possible to the future base pipe. Our analysis and testing suggests that the new version should be a drop in replacement, but we'd really like you to try it out and give us some feedback before we submit to CRAN. You can install the development version from GitHub with:
```

```{r}
# Piped
warning("foo") %>% suppressWarnings()

# Nested
suppressWarnings(warning("foo"))
```

Thanks to this change, you will now be able to pipe into testthat error expectations, for instance:

```{r}
library(testthat) %>%
  suppressMessages()

{ 1 + "a" } %>%
  expect_error("non-numeric argument")
```

Note that one consequence of having a lazy pipe is that the whole pipeline will be shown on the call stack before any errors are thrown:

```{r, eval = FALSE}
f1 <- function(x) x
f2 <- function(x) x
f3 <- function(x) x
f4 <- function(x) x

stop("oh no") %>% f1() %>% f2() %>% f3() %>% f4()
#> Error in f1(.) : oh no

rlang::last_trace()
#> <error/rlang_error>
#> oh no
#> Backtrace:
#>     █
#>  1. ├─stop("oh no") %>% f1() %>% f2() %>% f3() %>% f4()
#>  2. ├─global::f4(.)
#>  3. ├─global::f3(.)
#>  4. ├─global::f2(.)
#>  5. └─global::f1(.)
```

The last function of the pipeline is `f4()`, so that's the first one to be run. It evaluates its argument which is provided by `f3()`, so that's the second function pushed on the stack. And so on until `f1()` needs the result of `stop("oh no")` which causes an error.

## Towards a release

> Though we have changed the behaviour of the pipe, none of the 2600 reverse dependencies of magrittr, purrr, tidyr, and dplyr were broken by the change. To be extra sure, we'd be grateful for any additional testing on real-life scripts with this development version.

Reviewer: Mention that we analysed theoretically too.
